File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class pair{
        boolean bsp=true;
        long min=Long.MAX_VALUE;
        long max=Long.MIN_VALUE;
    }
    public boolean isValidBST(TreeNode root) {
        return validate(root).bsp;
    }
    public pair validate(TreeNode root){
        if(root==null){
            return new pair();
        }
        pair l=validate(root.left);
        pair r=validate(root.right);
        pair lr=new pair();
        lr.min=Math.min(l.min,Math.min(root.val,r.min));
        lr.max=Math.max(l.max,Math.max(root.val,r.max));
        lr.bsp=l.bsp && r.bsp && l.max<root.val && r.min>root.val;
        return lr;
    }
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    class Pair{
    public:
       bool bsp;
       long long max;
       long long min;
       Pair(){
        bsp=true;
        max=LLONG_MIN;
        min=LLONG_MAX;
       } 
    };
    bool isValidBST(TreeNode* root) {
        return bst(root).bsp;
    }
    Pair bst(TreeNode* root){
        if(root==nullptr){
            return Pair();
        }
        Pair l=bst(root->left);
        Pair r=bst(root->right);
        Pair rp;
        rp.min=min(l.min,min((long long)root->val,r.min));
        rp.max=max(l.max,max((long long)root->val,r.max));
        rp.bsp=l.bsp && r.bsp && (l.max<root->val) && (r.min>root->val);
        return rp;
    }
};